Add prepro code (#17)
neuroinformatics-unit/course-ephys-osss: book/preprocessing-steps-coding.qmd

- **common median referencing**: Removes noise that takes the form of stripes through the data (i.e. occurring on all channels for a short period of time). This is applied across all channels, separately per time point.

- **highpass spatial filter**: A destriping algorithm (i.e. similar to common median referencing) introduced by the [International Brain Laboratory (IBL)](https://figshare.com/articles/online_resource/Spike_sorting_pipeline_for_the_International_Brain_Laboratory/19705522?file=49783080) that accounts for slow oscillations of the signal across channels within the noise stripe.

- **motion correction** : A step to account for physical drift of the probe in the brain. Attempts to estimate the drift and interpolate the channels (having the effect of shifting the signal up or down the probe) to account for it.

- **whiten** : Removes correlated noise across channels and helps isolate spikes on  a subset of channels. Has significant effects on the data scaling and waveform shape.

> 

Reviewer: Maybe worth saying something like "You don't apply all of these steps to your recording recordings. Which preprocessing steps you choose will depend on your data, e.g. your recording device, whether your implant is chronic or acute."

Just don't wanna give the impression that you should do all of these!

Author: thanks added as a warning as this is a critical point!

import spikeinterface.preprocessing as si_prepro
import spikeinterface.widgets as si_widgets
import matplotlib.pyplot as plt
```

# Downloading the data

TODO: include some information on the experiment, probably do this on a central 'dataset' page

First we need to download the dataset we will. This dataset is (TODO: brief introduction). For more information on the dataset and how to download any one of the thousands of open-access datasets available on the DANDI archive, please see TODO.

```{python}

LOCAL_FOLDER = Path("_tmp/nwb_first_slice")

if LOCAL_FOLDER.exists():
    # if we have already downloaded the data, reload it.
    recording = si.load(LOCAL_FOLDER)
else:
    # Otherwise, download the first 90 seconds from DANDI
    download_and_slice_nwb_from_dandi(
        s3_url="https://dandiarchive.s3.amazonaws.com/blobs/a8f/800/a8f8003e-4483-4b50-8a45-91ac5971f5d5",
        duration_s=90,
        local_folder=LOCAL_FOLDER
    )
    recording = si.load(LOCAL_FOLDER)

```

# Initial steps

Next, we will define a function to plot and compare multiple preprocessing steps. We will reuse this function through the rest of the tutorial.

```{python}

> def plot_prepro(recordings: dict, time_range: tuple[float]) -> None:

Reviewer: ```suggestion
def plot_preprocessing(recordings: dict, time_range: tuple[float]) -> None:
```

I would also remove the typing - I think it's confusing for new users. But that's controversial!

Author: agree!

    """
    Plot multiple recordings in subplots.

    Parameters
    ----------
    recordings :
        A dictionary of recordings. The recordings will be plot and
        the corresponding keys used as plot titles.
    time_range: 
        A tuple with the (start, stop) times of the data to plot.
    """
    fig, axes = plt.subplots(
        1, len(recordings), figsize=(6 * len(recordings), 5),
        squeeze=False, layout="constrained",
    )

>     for ax, (name, recording) in zip(axes.ravel(), recordings.items()):

Reviewer: This is very gross. Do you really need to ravel?

Author: That's what happens when you ask claude to write you a function and forget to review it! new version below

```
def plot_prepro(recordings, time_range):
    """
    Plot multiple recordings in subplots.

    Parameters
    ----------
    recordings :
        A dictionary of recordings. The recordings will be plot and
        the corresponding keys used as plot titles.
    time_range: 
        A tuple with the (start, stop) times of the data to plot.
    """
    fig, axes = plt.subplots(
        1, len(recordings), 
        figsize=(6 * len(recordings), 5), 
        squeeze=False,
        layout="constrained",
    )

    # Grab the first (and only) row of axes
    axes = axes[0, :]

    for idx, (name, recording) in enumerate(recordings.items()):
        
        ax = axes[idx]

        si_widgets.plot_traces(
            recording,
            time_range=time_range,
            mode="map",  # plot as a heatmap
            ax=ax,
            with_colorbar=True,
        )
        ax.set_title(name)
        ax.tick_params(axis="x", rotation=45)

```

        si_widgets.plot_traces(
            recording,
            time_range=time_range,
            mode="map",  # plot as a heatmap
            ax=ax,
            with_colorbar=True,
        )
        ax.set_title(name)
        ax.tick_params(axis="x", rotation=45)

```

First, let's have a look at the effect of bandpass filtering the data, using the [`bandpass_filter()`](https://spikeinterface.readthedocs.io/en/latest/api.html#spikeinterface.preprocessing.bandpass_filter) function. We will plot the first second of the data:

```{python}

import spikeinterface.preprocessing as si_prepro

recording_bp = si_prepro.bandpass_filter(
    recording, 

>     # commonly used values used based on AP waveform kinetics

Reviewer: ```suggestion
    # commonly used values used to find spiking events
```

or something. Waveform kinetics sounds a bit complicated!

)
```

**It didn't work!** We can still see the dead channels on our recording. Why is this? 

The IBL algorithm was developed for Neurpixel probes (with 384 recording channels) but we are using Cambridge Neurotech 64-channel probes. 
As such, the default settings are not appropriate for our data, and we need to adjust them slightly. We reduce the threshold for
dead channel detection from `-0.5` to `-0.25` (this was done through manual trial and error):

```{python}
# Let's try again with different settings
bad_channel_ids, channel_labels = si_prepro.detect_bad_channels(
    recording_bp, dead_channel_threshold=-0.25
)
print("second attempt labels:\n", channel_labels)

rec_clean = recording_bp.remove_channels(bad_channel_ids)

plot_prepro(
    {"filtered": recording_bp, 
    "bad channels removed": rec_clean},
    TIME_RANGE
)

```

Now the dead channels, clearly identifiable by visual inspection, are properly detected by the algorithm.

::: {.callout-warning}
This example demonstrates why it is critically important to **always visualize your data**
to ensure the default arguments of the algorithms we are using are suitable for the recording.
:::

# Common average referencing

> Electrophysiological recordings are often contaminated by noise artefacts that appear on all channels at once for a short period of time. These appear as vertical 'stripes' on the heatmap view of the data, as can be seen below. 

Reviewer: What are these actually due to, physically?

Author: good q added: "Striping artefacts are typically caused by electrical noise that is shared across many recording channels, for example fluctuations in the reference electrode or electromagnetic interference,"

Common average referencing (CAR) takes the average signal across all of the channels, and subtracts it. Sometimes you wil see this called 'common median referencing' (CMR) when the median is used instead of the mean. 

We can use SpikeInterface's [`common_reference()`](https://spikeinterface.readthedocs.io/en/latest/api.html#spikeinterface.preprocessing.common_reference) function to perform common median referencing: 

```{python}

recording_cmr = si_prepro.common_reference(
    rec_clean, 
    operator="median"
)

plot_prepro(
    {"filtered": rec_clean, 
    "recording_cmr": recording_cmr}, 
    TIME_RANGE
)

```

We can see that contaminating noise spikes that run vertically across the heatmap are removed through this processing method.

> The implicit assumption of CAR or CMR is that the variability in the noise stripe is fairly small across channels (and so can be approximated well by removing only the central tendency measure). However, in practice this assumption is often broken, with stripes appearing only on a subset of channels (e.g. the top half). As such, the `common_reference` function contains a number of arguments allowing the operation to be performed on subsets of channels, and these can be explored in the 'things to try' section below.

Reviewer: ```suggestion
The implicit assumption of CAR or CMR is that the variability in the noise stripe is fairly small across channels (and so can be approximated well by removing only the central tendency measure). However, in practice this assumption is often broken, with stripes appearing only on a subset of channels (e.g. the top half), especially for long (>1mm) probes. As such, the `common_reference` function contains a number of arguments allowing the operation to be performed on subsets of channels and these can be explored in the 'things to try' section below.
```

maybe

check with Olivier, maybe remove this section. Will need to adjust the parameters for the 64-channel probe
vs. NP1 probe it was designed for. We need to calculate physical probe length and compute cycles/um and
match these across probes. I did this and the Wn was too low for this probe so it might not be possible.

:::{.callout-tip}
### Things to try

There's no need to get through all of these — pick whatever looks interesting.

- The time window we plot is important. Each spike is around 1-5 ms long, so if we plot long time windows we don't have enough pixels on the screen to resolve spikes. Nonetheless, longer windows can reveal other patterns in the data (e.g. 50 Hz noise). Play around with plotting different time windows (e.g. 10 ms, 200 ms, 1 s) to get a feel for this.
- Plot the data using the `mode="line"` argument. Same data — a completely different perspective!
- What happens if you set the bandpass filter minimum cutoff to zero? What frequencies are including now? Or set the maximum cutoff to 15000?
- Try common median referencing (CMR) particular subsets of channels, using the `reference`, `groups` arguments as well

> as use different channels as reference using the `ref_channel_ids` argument. Recordings have a `get_channel_ids()` function hat

Reviewer: ```suggestion
as use different channels as reference using the `ref_channel_ids` argument. Recordings have a `get_channel_ids()` function that
```

Function hat!

returns a list of ids.

> - Compute the Fourier transform of a data before and after bandpass filtering (using the `.get_traces()` function and [scipy's](https://scipy.org/) fourier transform functions) and plot the frequency spectrum.

Reviewer: Have you tried this? I think it's a mess: scipy's default gives you a complex representation (as in: complex numbers). Much easier with pynapple (https://pynapple.org/user_guide/10_power_spectral_density.html) but a bit tricky at this stage...

Author: Yea previously I have left this in smaller courses for power-users who are likely to storm through the other questions, but it is unnecessarily open-ended. Maybe we could give a bit more detail e.g. to try `rfft`, `rfftfreqs` and take the magnitude of the output of `rfft`? Or make it clear this is 'hard'?

:::

# Motion correction

> 

Reviewer: I want to discuss this more generally with everyone. My understanding is that all the algorithms were designed for acutely inserted probes, and can exaggerate existing aretefacts for chronic recordings. I might try and make something about this...

Author: yea this would be very valuable I do worry about this step, and people not always checking the driftmaps. Actually maybe I can plug [driftplots](https://driftplots.neuroinformatics.dev/) below

The recording probe can move during an experiment. The major axis of movement is typically the direction along the probe's insertion path, as this offers less physical resistance than moving into the surrounding brain. Movement along this axis has the effect of shifting the recorded waveforms up or down the probe.

The majority of motion correction algorithms are based around the movement of detected spike waveforms. The pattern of detected spikes is used as a 'barcode' across time, allowing motion to be tracked through the recording (see this 

> [excellent infographic](https://spikeinterface.readthedocs.io/en/stable/_images/motion_correction_overview.png) for details). Therefore most motion correction algorithms have three main steps:

Reviewer: ```suggestion
[infographic](https://spikeinterface.readthedocs.io/en/stable/_images/motion_correction_overview.png) for details - the grey dots in the plots are individual spikes which have been detected). Most motion correction algorithms have three main steps:
```

Don't wannt self-praise too much! Also worth describing what's in it.

    preset="dredge_fast",
    output_motion_info=True,
    # turn off non-rigid estimation
    estimate_motion_kwargs={"rigid": True},  
)
print(rec_corrected)

```

An [`InterpolateMotionRecording`](https://github.com/SpikeInterface/spikeinterface/blob/98170502207588f624a7dc88033ebf6f11a929cb/src/spikeinterface/sortingcomponents/motion/motion_interpolation.py#L274) is returned. This recording object applies motion correction to the requested chunk of data
based on the estimated motion.

Once motion correction is run, we can plot the outputs to inspect the results:

```{python}

# Display the motion output: drift map + estimated motion over depth/time.
fig = plt.figure(figsize=(14, 8))
si_widgets.plot_motion_info(
    motion_info,
    recording=rec_corrected,
    figure=fig,
    color_amplitude=True,
    amplitude_cmap="inferno",
    scatter_decimate=10,
)
fig.suptitle("Motion correction output")

```

In the top two plots, we see a spike raster plot before (left) and after (right) motion correction. Each dot indicates a spike, with its depth on the y-axis and recording time on the x-axis. 

The bottom left plot shows the estimated physical motion in μm over time.

> Because this is only a 90-second recording, we see very little drift. On longer recordings, the drift and its correction will be more apparent, for example on these example recording with injected drift (from [this useful explainer](https://spikeinterface.readthedocs.io/en/stable/how_to/handle_drift.html#handle-drift-in-your-recording)):

Reviewer: I'll do a full 2 hour plot for this recording!

Reviewer: <img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/df9558eb-2c66-478a-8372-27de098b27f4" />

Author: Nice, thanks!! This is a good example of when not to apply drift correction

In SpikeInterface, whitening is done by the [`whiten()`](https://spikeinterface.readthedocs.io/en/latest/api.html#spikeinterface.preprocessing.whiten) function:

```{python}

recording_whiten = si_prepro.whiten(rec_corrected)

plot_prepro(
    {"rec_corrected": rec_corrected, 
    "whitened": recording_whiten}, 
    TIME_RANGE
)

```

This is all we need to do to get a whitened version of our recording, however due the scaling that whitening performs it is difficult to see what has happened. 

To get a better understanding of how whitening changes the data, let's rescale the whitened data back to its original scaling (roughly) for visualization purposes. **Note we would not ever do this for recordings used in downstream analysis, this is just for visualization purposes**.

::: {.callout-note collapse="true"}
### How the rescaling works.

The correlation matrix is the key structure of the whitening step. It is computed by correlating all channels with each other:

::: {#fig-kilosort-like fig-align="center"}
![](images/whiten-mat.png){width=100%}
:::

> Here $c_i$ represents a channels data indexed by some integer $i$. $c_ic_j$ represents the sum of squares of the data from channels $i$ and $j$ (i.e. multiple each value in the vector with the corresponding value and add these all together). For example $c_1c_1$ is the variance for channel $c_1$, and $c_1c_2$ represents the covariance between channels 1 and 2. 

Reviewer: ```suggestion
Here $c_i$ represents a channel's data indexed by some integer $i$. $c_ic_j$ represents the sum of squares of the data from channels $i$ and $j$ (i.e. multiple each value in the vector with the corresponding value and add these all together). For example $c_1c_1$ is the variance for channel $c_1$, and $c_1c_2$ represents the covariance between channels 1 and 2. 
```

For those interested, you can see how this comes from the [covariance definition](https://www.educba.com/covariance-formula/) when the mean of the channel data is zero and we ignore the scaling factor $N$.

The goal of the whitening step is to remove all cross-channel correlation from the data. As we can see from the above matrix, this equates to setting all the off-diagonal entries to zero. In fact it goes a little further, setting the variance of each channel to `1` as well (making the covariance matrix the 'identity' matrix):

::: {#fig-kilosort-like fig-align="center"}
![](images/whiten-mat-identity.png){width=100%}
:::

Rescaling the variances is why this step has such a large effect on the data scaling.

> We can estimate the original channel scaling by multiplying the data by the average standard deviation (i.e. noise) of the data. This scales back up the variances along the covariance matrix to approximate that in the original dataset.

Reviewer: I find the second sentence "This scales..." very confusing.

Author: yes am having trouble expressing this, I  tried:

"We can estimate the original channel scaling by multiplying the data by the average standard deviation (i.e. noise) of the data.
When we compute the covariance matrix of the scaled data, we should see this approximation of the original channel
variances along the diagonal, instead of ones (with zeros still on the off-diagonals, at least in theory)."

but maybe its easier just to leave it as the first sentence?

```{python}

# Select the times we want to display and convert them into samples
start_frame = int(TIME_RANGE[0] * rec_corrected.get_sampling_frequency())
end_frame = int(TIME_RANGE[1] * rec_corrected.get_sampling_frequency())

# Get the data from this time range, and further slice it around a few spikes
orig_data = rec_corrected.get_traces(
    start_frame=start_frame, end_frame=end_frame, return_in_uV=True
)[200:400, :4]

whitened_data = scaled_white_recording.get_traces(
    start_frame=start_frame, end_frame=end_frame, return_in_uV=True,
)[200:400, :4]

# Plot the spikes, adding an offset between traces so they 
# are displayed vertically stacked.
plt.figure(figsize=(8, 5))
plt.plot(orig_data + np.arange(orig_data.shape[1]) * 750, color="k")
plt.plot(whitened_data + np.arange(whitened_data.shape[1]) * 750, color="r")
plt.xlabel("sample number")
plt.show()

```

Here, we can see that spikes are 'squashed' onto a smaller number of channels (which can aid spike sorting). The second spike is now only found on 4 channels instead of 3. The 1st and 3rd spikes, which was probably a single waveform distributed over 4 channels, however now appear to be split into two separate spikes.

> # Turn off the steps in KS! 

Reviewer: ```suggestion
# How to turn off the steps in KS 
```

Original sounds aggressively anti-KS!
